File: src/nodejs/mapper.js
/*
 * Mapper function handler
 *
 * */

var async = require('async');
var AWS = require('aws-sdk');
var zlib = require('zlib');
var es = require('event-stream');
var s3 = new AWS.S3();
var utils = require('./utils');
utils = new utils();

var procUtil = require('util');
var TASK_MAPPER_PREFIX = "task/mapper/";

exports.handler = function(event, context) {
    
    // calc execution time 
    var startTime = process.hrtime();
    
    var jobBucket = event.jobBucket;
    var srcBucket = event.bucket;
    var srcKeys = event.keys;
    var jobId = event.jobId;
    var mapperId = event.mapperId;
    
    // Counters
    var lineCount = 0;
    var totals = {};

    // CSV Processor to aggregate ad revenue for IPs 
    var mapfunc = function(srcKey, cb){
        //console.log("processing", srcKey);
        try {      
            var pt = 0; // processing time
           
            // Create our processor that splits lines/words
            var lineSplitter = es.split();
            lineSplitter.on('data', function(line){
                var dataProcessingStartTime = process.hrtime(); 
                lineCount++;
                data = line.split(',');
                srcIp = data[0].substr(1, 8);
                if(!totals[srcIp])
                    totals[srcIp] = 0;
                totals[srcIp] += parseFloat(data[3]);

                var et = process.hrtime(dataProcessingStartTime);
                pt += (et[0] * 1e9 + et[1]);
            })
            .on('error', function(err){
                console.error('Error in lineSplitter: ' + err);
            })
            .on('end', function(err){
                //console.log("End processing", procUtil.inspect(process.memoryUsage()));
                if (err){
                    console.error(err);
                    cb(err, null);
                    return;
                }
                cb(null, [srcKey, pt]);
            });

            // console.log('Downloading key ' + srcKey);
            
            ///// Create our pipeline /////
            //
            // If input files are gzipped
            // TODO: perhaps check content type as well?
            if (srcKey.match(/\.gz/)){
                s3.getObject({
                    Bucket: srcBucket,
                    Key: srcKey
                })
                .createReadStream()
                    .pipe(zlib.createGunzip())
                    .pipe(lineSplitter);
            }
            else {
                s3.getObject({
                    Bucket: srcBucket,
                    Key: srcKey
                })
                .createReadStream()
                .pipe(lineSplitter);
            }
        }catch (e){
            // Catch all during map processing
            console.error(e);
            cb(e, null);
        }
    };

    // Exec in Parallel
    async.map(srcKeys, mapfunc, function (err, results) {
        var tDiff = process.hrtime(startTime);
        var s3DownloadTime = tDiff[0] * 1e9 + tDiff[1];
        if (err){
            console.error(err);
            context.fail(err);
        }
        var totalProcessingTime = 0;  // Data processing time
        for (var i = 0, len = results.length; i < len; i++){
            totalProcessingTime +=  results[i][2];
        }
        var tDiff = process.hrtime(startTime);
        var timeTaken = tDiff[0] * 1e9 + tDiff[1];
        var timeInSecs = timeTaken/1000000000;

        // S3 File download time
        s3DownloadTime -= totalProcessingTime;

        //  S3Keys, TotalLines, TotalTime, S3DownloadTime, TotalDataProcessingTime
        pret = [srcKeys.length, lineCount, timeTaken, s3DownloadTime/1000000000, totalProcessingTime/1000000000]; //for print
        //console.log(pret);

        // Write mapper output
        var mapperFname = jobId + "/" + TASK_MAPPER_PREFIX + mapperId; 
        var metadata = {
                            linecount: '' + lineCount,
                            processingtime: '' + timeInSecs,
                            memoryUsage: JSON.stringify(process.memoryUsage())
                        };
        //console.log("before upload", procUtil.inspect(process.memoryUsage()));
        utils.writeToS3(s3, jobBucket, mapperFname, utils.objToCSVBuffer(totals), metadata, function(err, data){
            //console.log("after upload", procUtil.inspect(process.memoryUsage()));
            if(err)
                console.error("Write to S3 failed for " + mapperFname);
            context.succeed(pret);
        });
    });
};


/*
   var ev = {
   "bucket": "smallya-useast-1", 
   "keys": ["pavlo.sample"],
   "jobId": "jobid134",
   "mapperId": 1,
   "jobBucket": "smallya-useast-1"
   }
   var cx = {} 
   cx.succeed = function(x) { console.log(x);}
   exports.handler(ev,cx)
*/

File: src/python/mapper.py
'''
Python mapper function

Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: MIT-0
'''

import boto3
import json
import random
import resource
from io import StringIO
import time

# create an S3 session
s3 = boto3.resource('s3')
s3_client = boto3.client('s3')

# constants
TASK_MAPPER_PREFIX = "task/mapper/";

def write_to_s3(bucket, key, data, metadata):
    s3.Bucket(bucket).put_object(Key=key, Body=data, Metadata=metadata)

def lambda_handler(event, context):
    
    start_time = time.time()

    job_bucket = event['jobBucket']
    src_bucket = event['bucket']
    src_keys = event['keys']
    job_id = event['jobId']
    mapper_id = event['mapperId']
   
    # aggr 
    output = {}
    line_count = 0
    err = ''

    # INPUT CSV => OUTPUT JSON

    # Download and process all keys
    for key in src_keys:
        response = s3_client.get_object(Bucket=src_bucket,Key=key)
        contents = response['Body'].read()
        
        for line in contents.split('\n')[:-1]:
            line_count +=1
            try:
                data = line.split(',')
                srcIp = data[0][:8]
                if srcIp not in output:
                    output[srcIp] = 0
                output[srcIp] += float(data[3])
            except Exception, e:
                print(e)
                #err += '%s' % e

    time_in_secs = (time.time() - start_time)
    #timeTaken = time_in_secs * 1000000000 # in 10^9 
    #s3DownloadTime = 0
    #totalProcessingTime = 0 
    pret = [len(src_keys), line_count, time_in_secs, err]
    mapper_fname = "%s/%s%s" % (job_id, TASK_MAPPER_PREFIX, mapper_id) 
    metadata = {
                    "linecount":  '%s' % line_count,
                    "processingtime": '%s' % time_in_secs,
                    "memoryUsage": '%s' % resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
               }

    print("metadata", metadata)
    write_to_s3(job_bucket, mapper_fname, json.dumps(output), metadata)
    return pret

'''
ev = {
   "bucket": "-useast-1", 
   "keys": ["key.sample"],
   "jobId": "pyjob",
   "mapperId": 1,
   "jobBucket": "-useast-1"
   }
lambda_handler(ev, {});
'''
